runtime/zlib.cpp

```diff
@@ -58,13 +65,147 @@ const string_buffer *zlib_encode(const char *s, int32_t s_len, int32_t level, in
+    } else if (option.get_string_key() == string("window")) {
+      value = options.get_value(string("window"));
+      if (value.is_int() && value.as_int() >= 8 && value.as_int() <= 15) {
+        level = value.as_int();
+      } else {
+        php_warning("deflate_init() : option window should be number between 8..15");
+        return {};
+      }
+    } else if (option.get_string_key() == string("strategy")) {
+      value = options.get_value(string("strategy"));
+      if (value.is_int()) {
+        switch (value.as_int()) {
+          case Z_FILTERED:
+          case Z_HUFFMAN_ONLY:
+          case Z_RLE:
+          case Z_FIXED:
+          case Z_DEFAULT_STRATEGY:
+            strategy = value.as_int();
+            break;
+          default:
+            php_warning("deflate_init() : option strategy should be one of ZLIB_FILTERED, ZLIB_HUFFMAN_ONLY, ZLIB_RLE, ZLIB_FIXED or ZLIB_DEFAULT_STRATEGY");
+            return {};
+        }
+      } else {
+        php_warning("deflate_init() : option strategy should be one of ZLIB_FILTERED, ZLIB_HUFFMAN_ONLY, ZLIB_RLE, ZLIB_FIXED or ZLIB_DEFAULT_STRATEGY");
+        return {};
+      }
+    } else if (option.get_string_key() == string("dictionary")) {
+      php_warning("deflate_init() : option dictionary isn't supported yet");
+      return {};
+    } else {
+      php_warning("deflate_init() : unknown option name \"%s\"", option.get_string_key().c_str());
+      return {};
+    }
+  }
+
+  class_instance<C$DeflateContext> context;
+  context.alloc();
+
+  z_stream *stream = &context.get()->stream;
+  stream->zalloc = zlib_dynamic_alloc;
+  stream->zfree = zlib_dynamic_free;
+  stream->opaque = nullptr;
+
+  if (encoding < 0) {
+    encoding += 15 - window;
+  } else {
+    encoding -= 15 - window;
+  }
+
+  dl::CriticalSectionGuard guard;
+  int err = deflateInit2(stream, level, Z_DEFLATED, encoding, memory, strategy);
+  if (err != Z_OK) {
+    php_warning("deflate_init() : zlib error %s", zError(err));
+    context.destroy();
+    return {};
+  }
+  return context;
+}
+
+Optional<string> f$deflate_add(const class_instance<C$DeflateContext> &context, string &data, int64_t flush_type) {
+  switch (flush_type) {
+    case Z_BLOCK:
+    case Z_NO_FLUSH:
+    case Z_PARTIAL_FLUSH:
+    case Z_SYNC_FLUSH:
+    case Z_FULL_FLUSH:
+    case Z_FINISH:
+      break;
+    default:
+      php_warning("deflate_add() : flush type should be one of ZLIB_NO_FLUSH, ZLIB_PARTIAL_FLUSH, ZLIB_SYNC_FLUSH, ZLIB_FULL_FLUSH, ZLIB_FINISH, ZLIB_BLOCK, ZLIB_TREES");
+      return {};
+  }
+
+  string_buffer buffer;
+  int out_size = compressBound(data.size()) + 30;
+  buffer.reserve(out_size);
+  z_stream *stream = &context.get()->stream;
+  stream->next_in = reinterpret_cast<Bytef *>(data.buffer());
+  stream->next_out = reinterpret_cast<Bytef *>(buffer.buffer());
+  stream->avail_in = data.size();
+  stream->avail_out = out_size;
+
+  dl::CriticalSectionGuard guard;
+  int res = deflate(stream, flush_type);
```

Comment:
Are you sure we don't need to do it in a loop until `stream->avail_out = 0`?
Check this too: https://www.zlib.net/zlib_how.html